File: web/src/lib/__tests__/data-overview-access.test.ts
import assert from 'node:assert/strict'
import { describe, test } from 'node:test'

import {
  canAccessDataOverview,
  ROLE,
  type DataOverviewAccessUser,
} from '../roles'

describe('data overview entry access', () => {
  test('admins and root always pass', () => {
    assert.equal(canAccessDataOverview({ role: ROLE.ADMIN }), true)
    assert.equal(canAccessDataOverview({ role: ROLE.SUPER_ADMIN }), true)
  })

  test('BP role passes only with configured visible departments', () => {
    assert.equal(
      canAccessDataOverview({ role: ROLE.BU_BP, overview_dept_ids: ['dept:1:team'] }),
      true
    )
    assert.equal(
      canAccessDataOverview({ role: ROLE.BU_BP, overview_dept_ids: ['dept:1:team', 'dept:1:bu'] }),
      true
    )
    assert.equal(canAccessDataOverview({ role: ROLE.BU_BP, overview_dept_ids: [] }), false)
    assert.equal(canAccessDataOverview({ role: ROLE.BU_BP }), false)
  })

  test('BP without configured departments is denied even when leading departments', () => {
    assert.equal(
      canAccessDataOverview({ role: ROLE.BU_BP, overview_dept_ids: [], is_dept_leader: true }),
      false
    )
  })

  test('common users pass only as department leaders', () => {
    assert.equal(canAccessDataOverview({ role: ROLE.USER, is_dept_leader: true }), true)
    assert.equal(canAccessDataOverview({ role: ROLE.USER }), false)
  })

  test('null user is denied', () => {
    assert.equal(canAccessDataOverview(null as unknown as DataOverviewAccessUser), false)
    assert.equal(canAccessDataOverview(undefined), false)
  })
})

File: middleware/data_overview_access_test.go
package middleware

import (
	"testing"

	"github.com/QuantumNous/new-api/common"
	"github.com/QuantumNous/new-api/model"
	"github.com/stretchr/testify/assert"
)

func TestCanAccessDataOverview(t *testing.T) {
	tests := []struct {
		name     string
		user     *model.User
		expected bool
	}{
		{
			name:     "root can access",
			user:     &model.User{Role: common.RoleRootUser},
			expected: true,
		},
		{
			name:     "admin can access",
			user:     &model.User{Role: common.RoleAdminUser},
			expected: true,
		},
		{
			name:     "BP with configured bp_level can access",
			user:     &model.User{Role: common.RoleBUBP, BpLevel: 2},
			expected: true,
		},
		{
			name:     "BP with bp_level 1 can access",
			user:     &model.User{Role: common.RoleBUBP, BpLevel: 1},
			expected: true,
		},
		{
			name:     "BP with bp_level 0 is denied",
			user:     &model.User{Role: common.RoleBUBP, BpLevel: 0},
			expected: false,
		},
		{
			name:     "BP with bp_level 0 is denied even when leading departments",
			user:     &model.User{Role: common.RoleBUBP, BpLevel: 0, OpenId: "ou_bp", Departments: `[{"department_id":"team","leaders":[{"leader_id":"ou_bp"}]}]`},
			expected: false,
		},
		{
			name:     "dept leader with common role can access",
			user:     &model.User{Role: common.RoleCommonUser, OpenId: "ou_leader", Departments: `[{"department_id":"team","leaders":[{"leader_id":"ou_leader"}]}]`},
			expected: true,
		},
		{
			name:     "common user without leader departments is denied",
			user:     &model.User{Role: common.RoleCommonUser},
			expected: false,
		},
		{
			name:     "guest user is denied",
			user:     &model.User{Role: common.RoleGuestUser},
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.expected, canAccessDataOverview(tt.user))
		})
	}
}
